File: 프로그래머스/lv3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.java
import java.util.*;
import java.io.*;

class Solution {
    public int[] solution(String[] gems) {
        int[] answer = new int[2];
        int size = gems.length;
        Set<String> set = new HashSet<>();
        Map<String, Integer> map = new HashMap<>();
        for(String s : gems){
            set.add(s);
        }        
        int left = 0, right = 0;
        int dist = Integer.MAX_VALUE;
        while(true){
            if(set.size() == map.size()){
                map.put(gems[left], map.get(gems[left])-1);
                if(map.get(gems[left]) == 0) map.remove(gems[left]);
                left++;
            }
            else if(right == size) break;
            else{
                map.put(gems[right], map.getOrDefault(gems[right], 0) + 1);
                right++;
            }
            
            if(map.size() == set.size()){
                if(right - left < dist){
                    dist = right - left;
                    answer[0] = left + 1;
                    answer[1] = right;
                }
            }
        }
        return answer;
    }
}

File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.cpp
#include <string>
#include <vector>
#include <set>
#include <map>
using namespace std;

vector<int> solution(vector<string> gems) {
    vector<int> answer(2);
    set<string> s;
    for (auto g : gems) {
        s.insert(g);
    }
    int cnt = s.size();
    
    int st = 0;
    int ed = 0;
    int dist = 100001;
    map<string, int> m;
    
    
    while (st <= ed && ed < gems.size()) {
        m[gems[ed]]++;
        
        while (m.size() == cnt) {
            if (ed - st + 1 < dist) {
                dist = ed - st + 1;
                answer[0] = st + 1;
                answer[1] = ed + 1;
            }
            m[gems[st]]--;
            if (m[gems[st]] <= 0) m.erase(gems[st]);
            st++;
        }
        ed++;
    }
    return answer;
}